File: sdks/go/bridges/conformance_bridge.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"os"

	"github.com/autonoma-ai/sdk/sdks/go/v2/autonoma"
)

type bridgeInput struct {
	Module   string         `json:"module"`
	Function string         `json:"function"`
	Input    map[string]any `json:"input"`
}

type bridgeOutput struct {
	OK        bool   `json:"ok"`
	Result    any    `json:"result"`
	HasResult bool   `json:"-"`
	Error     string `json:"error,omitempty"`
}

func (b bridgeOutput) MarshalJSON() ([]byte, error) {
	if !b.OK {
		return json.Marshal(struct {
			OK    bool   `json:"ok"`
			Error string `json:"error,omitempty"`
		}{OK: false, Error: b.Error})
	}
	return json.Marshal(struct {
		OK     bool `json:"ok"`
		Result any  `json:"result"`
	}{OK: true, Result: b.Result})
}

func main() {
	data, err := io.ReadAll(os.Stdin)
	if err != nil {
		writeError(fmt.Sprintf("failed to read stdin: %v", err))
		return
	}

	var input bridgeInput
	if err := json.Unmarshal(data, &input); err != nil {
		writeError(fmt.Sprintf("failed to parse input: %v", err))
		return
	}

	dispatch(input)
}

func dispatch(input bridgeInput) {
	key := input.Module + "." + input.Function

	defer func() {
		if r := recover(); r != nil {
			writeError(fmt.Sprintf("%v", r))
		}
	}()

	switch key {
	case "hmac.signBody":
		body, _ := input.Input["body"].(string)
		secret, _ := input.Input["secret"].(string)
		writeResult(autonoma.SignBody(body, secret))

	case "hmac.verifySignature":
		body, _ := input.Input["body"].(string)
		signature, _ := input.Input["signature"].(string)
		secret, _ := input.Input["secret"].(string)
		writeResult(autonoma.VerifySignature(body, signature, secret))

	case "refs.signRefs":
		secret, _ := input.Input["secret"].(string)
		result, err := autonoma.SignRefs(toRefsPayload(input.Input["payload"]), secret)
		if err != nil {
			writeError(err.Error())
			return
		}
		writeResult(result)

	case "refs.verifyRefs":
		token, _ := input.Input["token"].(string)
		secret, _ := input.Input["secret"].(string)
		result, err := autonoma.VerifyRefs(token, secret)
		if err != nil {
			writeError(err.Error())
			return
		}
		writeResult(map[string]any{
			"refs":        result.Refs,
			"testRunId":   result.TestRunID,
			"environment": result.Environment,
		})

	case "unique.uniqueToken":
		testRunID, _ := input.Input["testRunId"].(string)
		writeResult(autonoma.UniqueToken(testRunID, toStringSlice(input.Input["parts"])...))

	case "unique.uniqueId":
		testRunID, _ := input.Input["testRunId"].(string)
		prefix, _ := input.Input["prefix"].(string)
		writeResult(autonoma.UniqueID(testRunID, prefix, toStringSlice(input.Input["parts"])...))

	case "unique.uniqueSlug":
		testRunID, _ := input.Input["testRunId"].(string)
		base, _ := input.Input["base"].(string)
		writeResult(autonoma.UniqueSlug(testRunID, base, toStringSlice(input.Input["parts"])...))

	case "unique.uniqueEmail":
		testRunID, _ := input.Input["testRunId"].(string)
		local, _ := input.Input["local"].(string)
		domain, _ := input.Input["domain"].(string)
		writeResult(autonoma.UniqueEmail(testRunID, local, domain))

	default:
		writeError(fmt.Sprintf("Unknown function: %s", key))
	}
}

func toStringSlice(v any) []string {
	arr, ok := v.([]any)
	if !ok {
		return nil
	}
	out := make([]string, 0, len(arr))
	for _, e := range arr {
		s, _ := e.(string)
		out = append(out, s)
	}
	return out
}

func writeResult(result any) {
	out := bridgeOutput{OK: true, Result: result, HasResult: true}
	data, _ := json.Marshal(out)
	fmt.Println(string(data))
}

func writeError(msg string) {
	out := bridgeOutput{OK: false, Error: msg}
	data, _ := json.Marshal(out)
	fmt.Println(string(data))
}

func toRefsPayload(v any) autonoma.RefsPayload {
	m, ok := v.(map[string]any)
	if !ok {
		return autonoma.RefsPayload{}
	}
	testRunID, _ := m["testRunId"].(string)
	environment, _ := m["environment"].(string)
	return autonoma.RefsPayload{
		Refs:        m["refs"],
		TestRunID:   testRunID,
		Environment: environment,
	}
}

File: sdks/rust/bridges/conformance_bridge.rs
//! Conformance test bridge for the Rust SDK.
//!
//! Reads a JSON test case from stdin, dispatches to the appropriate SDK
//! function, and writes the result to stdout. Scenario-v2 dropped the
//! create-graph interpreter and `fingerprint`; the Rust SDK conforms on the
//! version-agnostic `hmac`, `refs`, and `unique` primitives.

use serde_json::{json, Value};
use std::io::{self, Read};

use autonoma_sdk::hmac::{sign_body, verify_signature};
use autonoma_sdk::refs::{sign_refs, verify_refs, RefsPayload};
use autonoma_sdk::unique::{unique_email, unique_id, unique_slug, unique_token};

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();

    let data: Value = serde_json::from_str(&input).unwrap();

    let module = data["module"].as_str().unwrap_or("");
    let function = data["function"].as_str().unwrap_or("");
    let inp = &data["input"];

    match dispatch(module, function, inp) {
        Ok(val) => println!("{}", json!({ "ok": true, "result": val })),
        Err(e) => println!("{}", json!({ "ok": false, "error": e })),
    }
}

fn dispatch(module: &str, function: &str, inp: &Value) -> Result<Value, String> {
    match (module, function) {
        ("hmac", "signBody") => {
            let body = inp["body"].as_str().ok_or("missing body")?;
            let secret = inp["secret"].as_str().ok_or("missing secret")?;
            Ok(Value::String(sign_body(body, secret)))
        }
        ("hmac", "verifySignature") => {
            let body = inp["body"].as_str().ok_or("missing body")?;
            let signature = inp["signature"].as_str().ok_or("missing signature")?;
            let secret = inp["secret"].as_str().ok_or("missing secret")?;
            Ok(Value::Bool(verify_signature(body, signature, secret)))
        }
        ("refs", "signRefs") => {
            let payload: RefsPayload =
                serde_json::from_value(inp["payload"].clone()).map_err(|e| e.to_string())?;
            let secret = inp["secret"].as_str().ok_or("missing secret")?;
            Ok(Value::String(sign_refs(&payload, secret)))
        }
        ("refs", "verifyRefs") => {
            let token = inp["token"].as_str().ok_or("missing token")?;
            let secret = inp["secret"].as_str().ok_or("missing secret")?;
            let payload = verify_refs(token, secret)?;
            Ok(json!({
                "refs": payload.refs,
                "testRunId": payload.test_run_id,
                "environment": payload.environment,
            }))
        }
        ("unique", "uniqueToken") => {
            let test_run_id = inp["testRunId"].as_str().ok_or("missing testRunId")?;
            let parts: Vec<&str> = inp["parts"]
                .as_array()
                .map(|a| a.iter().filter_map(|v| v.as_str()).collect())
                .unwrap_or_default();
            Ok(Value::String(unique_token(test_run_id, &parts)))
        }
        ("unique", "uniqueId") => {
            let test_run_id = inp["testRunId"].as_str().ok_or("missing testRunId")?;
            let prefix = inp["prefix"].as_str().ok_or("missing prefix")?;
            let parts: Vec<&str> = inp["parts"]
                .as_array()
                .map(|a| a.iter().filter_map(|v| v.as_str()).collect())
                .unwrap_or_default();
            Ok(Value::String(unique_id(test_run_id, prefix, &parts)))
        }
        ("unique", "uniqueSlug") => {
            let test_run_id = inp["testRunId"].as_str().ok_or("missing testRunId")?;
            let base = inp["base"].as_str().ok_or("missing base")?;
            let parts: Vec<&str> = inp["parts"]
                .as_array()
                .map(|a| a.iter().filter_map(|v| v.as_str()).collect())
                .unwrap_or_default();
            Ok(Value::String(unique_slug(test_run_id, base, &parts)))
        }
        ("unique", "uniqueEmail") => {
            let test_run_id = inp["testRunId"].as_str().ok_or("missing testRunId")?;
            let local = inp["local"].as_str().ok_or("missing local")?;
            let domain = inp["domain"].as_str().ok_or("missing domain")?;
            Ok(Value::String(unique_email(test_run_id, local, domain)))
        }
        _ => Err(format!("Unknown function: {}.{}", module, function)),
    }
}
